source/validate_cfg.cpp

```diff
@@ -59,56 +60,97 @@ using bb_ptr = BasicBlock*;
 using cbb_ptr = const BasicBlock*;
 using bb_iter = vector<BasicBlock*>::const_iterator;
 
-/// @brief Sorts the blocks in a CFG given the entry node
+using get_blocks_func = function<const vector<BasicBlock*>*(const BasicBlock*)>;
+
+struct block_info {
+  cbb_ptr block;  ///< pointer to the block
+  bb_iter iter;   ///< Iterator to the current child node being processed
+};
+
+/// Returns true if a block with @p id is found in the @p work_list vector
 ///
-/// Returns a vector of basic block pointers in a Control Flow Graph(CFG) which
-/// are sorted in the order they were accessed in a post order traversal.
+/// @param[in] work_list Set of blocks visited in the the depth first traversal
+///                   of the CFG
+/// @param[in] id The ID of the block being checked
+/// @return true if the edge work_list.back().block->get_id() => id is a
+/// back-edge
+bool FindInWorkList(vector<block_info> work_list, uint32_t id) {
```

Comment:
Just noticed this makes a copy when it doesn't need to.
Should be a const ref to the vector.